File: src/mcp_server_check/tools/contractors.py
"""Contractor tools for the Check API."""

from __future__ import annotations

from fastmcp import FastMCP

from mcp_server_check.types import Address, FormParameter
from mcp_server_check.helpers import (
    Ctx,
    build_body,
    build_params,
    check_api_get,
    check_api_list,
    check_api_patch,
    check_api_post,
)


async def list_contractors(
    ctx: Ctx,
    company: str | None = None,
    limit: int | None = None,
    ids: list[str] | None = None,
    cursor: str | None = None,
) -> dict:
    """List contractors, optionally filtered by company.

    Args:
        company: Filter to contractors belonging to this Check company ID (e.g. "com_xxxxx").
        limit: Maximum number of results to return (default 10, max 100).
        ids: Filter to specific contractor IDs.
        cursor: Pagination cursor from a previous response.
    """
    return await check_api_list(
        ctx, "/contractors", params=build_params(company=company, limit=limit, ids=ids, cursor=cursor)
    )


async def get_contractor(ctx: Ctx, contractor_id: str) -> dict:
    """Get details for a specific contractor.

    Args:
        contractor_id: The Check contractor ID (e.g. "ctr_xxxxx").
    """
    return await check_api_get(ctx, f"/contractors/{contractor_id}")


async def create_contractor(
    ctx: Ctx,
    company: str,
    type: str | None = None,
    first_name: str | None = None,
    middle_name: str | None = None,
    last_name: str | None = None,
    business_name: str | None = None,
    dob: str | None = None,
    start_date: str | None = None,
    termination_date: str | None = None,
    workplaces: list[str] | None = None,
    primary_workplace: str | None = None,
    email: str | None = None,
    ssn: str | None = None,
    ein: str | None = None,
    default_net_pay_split: str | None = None,
    payment_method_preference: str | None = None,
    address: Address | None = None,
    metadata: str | None = None,
) -> dict:
    """Create a new contractor.

    Args:
        company: The Check company ID.
        type: Contractor type — "individual" or "business".
        first_name: Contractor's first name (or business primary contact's first name).
        middle_name: Contractor's middle name.
        last_name: Contractor's last name (or business primary contact's last name).
        business_name: Business name (for business-type contractors).
        dob: Date of birth (YYYY-MM-DD).
        start_date: Most recent start date of contract (YYYY-MM-DD).
        termination_date: Most recent termination date (YYYY-MM-DD).
        workplaces: List of workplace IDs where the contractor works.
        primary_workplace: Workplace ID of the contractor's primary workplace.
        email: Contractor's email address.
        ssn: Contractor's Social Security Number.
        ein: Contractor's Employer Identification Number (for businesses).
        default_net_pay_split: ID of contractor's default net pay split.
        payment_method_preference: "direct_deposit" or "manual".
        address: Address with keys: line1, line2, city, state, postal_code, country.
        metadata: Additional JSON metadata string.
    """
    return await check_api_post(
        ctx,
        "/contractors",
        data=build_body(
            {"company": company},
            type=type,
            first_name=first_name,
            middle_name=middle_name,
            last_name=last_name,
            business_name=business_name,
            dob=dob,
            start_date=start_date,
            termination_date=termination_date,
            workplaces=workplaces,
            primary_workplace=primary_workplace,
            email=email,
            ssn=ssn,
            ein=ein,
            default_net_pay_split=default_net_pay_split,
            payment_method_preference=payment_method_preference,
            address=address,
            metadata=metadata,
        ),
    )


async def update_contractor(
    ctx: Ctx,
    contractor_id: str,
    type: str | None = None,
    first_name: str | None = None,
    middle_name: str | None = None,
    last_name: str | None = None,
    business_name: str | None = None,
    dob: str | None = None,
    start_date: str | None = None,
    termination_date: str | None = None,
    workplaces: list[str] | None = None,
    primary_workplace: str | None = None,
    email: str | None = None,
    ssn: str | None = None,
    ein: str | None = None,
    default_net_pay_split: str | None = None,
    payment_method_preference: str | None = None,
    address: Address | None = None,
    metadata: str | None = None,
) -> dict:
    """Update an existing contractor.

    Args:
        contractor_id: The Check contractor ID.
        type: Contractor type — "individual" or "business".
        first_name: Contractor's first name.
        middle_name: Contractor's middle name.
        last_name: Contractor's last name.
        business_name: Business name (for business-type contractors).
        dob: Date of birth (YYYY-MM-DD).
        start_date: Most recent start date of contract (YYYY-MM-DD).
        termination_date: Most recent termination date (YYYY-MM-DD).
        workplaces: List of workplace IDs where the contractor works.
        primary_workplace: Workplace ID of the contractor's primary workplace.
        email: Contractor's email address.
        ssn: Contractor's Social Security Number.
        ein: Contractor's Employer Identification Number (for businesses).
        default_net_pay_split: ID of contractor's default net pay split.
        payment_method_preference: "direct_deposit" or "manual".
        address: Address with keys: line1, line2, city, state, postal_code, country.
        metadata: Additional JSON metadata string.
    """
    return await check_api_patch(
        ctx,
        f"/contractors/{contractor_id}",
        data=build_body(
            {},
            type=type,
            first_name=first_name,
            middle_name=middle_name,
            last_name=last_name,
            business_name=business_name,
            dob=dob,
            start_date=start_date,
            termination_date=termination_date,
            workplaces=workplaces,
            primary_workplace=primary_workplace,
            email=email,
            ssn=ssn,
            ein=ein,
            default_net_pay_split=default_net_pay_split,
            payment_method_preference=payment_method_preference,
            address=address,
            metadata=metadata,
        ),
    )


async def onboard_contractor(ctx: Ctx, contractor_id: str) -> dict:
    """Onboard a contractor, transitioning them to active status.

    Args:
        contractor_id: The Check contractor ID.
    """
    return await check_api_post(ctx, f"/contractors/{contractor_id}/onboard")


async def list_contractor_payments_for_contractor(
    ctx: Ctx,
    contractor_id: str,
    limit: int | None = None,
    cursor: str | None = None,
    payroll: str | None = None,
    status: str | None = None,
) -> dict:
    """List payments for a specific contractor.

    Args:
        contractor_id: The Check contractor ID.
        limit: Maximum number of results to return.
        cursor: Pagination cursor.
        payroll: Filter by payroll ID.
        status: Filter by status — "pending", "processing", "failed", or "paid".
    """
    return await check_api_list(
        ctx,
        f"/contractors/{contractor_id}/payments",
        params=build_params(limit=limit, cursor=cursor, payroll=payroll, status=status),
    )


async def get_contractor_payment_for_payroll(
    ctx: Ctx, contractor_id: str, payroll_id: str
) -> dict:
    """Get a contractor payment for a specific payroll.

    Args:
        contractor_id: The Check contractor ID.
        payroll_id: The Check payroll ID.
    """
    return await check_api_get(
        ctx, f"/contractors/{contractor_id}/payments/{payroll_id}"
    )


async def list_contractor_forms(
    ctx: Ctx,
    contractor_id: str,
    limit: int | None = None,
    cursor: str | None = None,
) -> dict:
    """List forms for a contractor.

    Args:
        contractor_id: The Check contractor ID.
        limit: Maximum number of results to return.
        cursor: Pagination cursor.
    """
    return await check_api_list(
        ctx,
        f"/contractors/{contractor_id}/forms",
        params=build_params(limit=limit, cursor=cursor),
    )


async def submit_contractor_form(
    ctx: Ctx,
    contractor_id: str,
    form_id: str,
    parameters: list[FormParameter] | None = None,
) -> dict:
    """Submit a contractor form.

    Args:
        contractor_id: The Check contractor ID.
        form_id: The form ID.
        parameters: List of name/value dicts representing form fields.
            Example: [{"name": "field_name", "value": "field_value"}].
    """
    body: dict | None = None
    if parameters is not None:
        body = {"parameters": parameters}
    return await check_api_post(
        ctx, f"/contractors/{contractor_id}/forms/{form_id}/submit", data=body
    )


def register(mcp: FastMCP, *, read_only: bool = False) -> None:
    mcp.add_tool(list_contractors)
    mcp.add_tool(get_contractor)
    mcp.add_tool(list_contractor_payments_for_contractor)
    mcp.add_tool(get_contractor_payment_for_payroll)
    mcp.add_tool(list_contractor_forms)
    if not read_only:
        mcp.add_tool(create_contractor)
        mcp.add_tool(update_contractor)
        mcp.add_tool(onboard_contractor)
        mcp.add_tool(submit_contractor_form)

File: ts/src/tools/contractors.ts
/** Contractor tools for the Check API. */

import { z } from "zod";
import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import {
  checkApiGet,
  checkApiPost,
  checkApiPatch,
  checkApiList,
  type CheckApiOptions,
} from "../helpers.js";
import type { ToolFilter } from "../tool-filter.js";
import { tool } from "./_register.js";



export function register(
  server: McpServer,
  api: CheckApiOptions,
  filter: ToolFilter,
  toolsetName: string,
) {
  tool(
    server,
    filter,
    toolsetName,
    "list_contractors",
    "List contractors, optionally filtered by company.",
    {
      company: z.string().optional().describe('Filter to contractors belonging to this Check company ID (e.g. "com_xxxxx").'),
      limit: z.number().optional().describe("Maximum number of results to return (default 10, max 100)."),
      ids: z.array(z.string()).optional().describe("Filter to specific contractor IDs."),
      cursor: z.string().optional().describe("Pagination cursor from a previous response."),
    },
    async ({ company, limit, ids, cursor }) => {
      const params: Record<string, unknown> = {};
      if (company !== undefined) params.company = company;
      if (limit !== undefined) params.limit = limit;
      if (ids !== undefined) params.ids = ids.join(",");
      if (cursor !== undefined) params.cursor = cursor;
      const result = await checkApiList(api, "/contractors", params);
      return { content: [{ type: "text" as const, text: JSON.stringify(result, null, 2) }] };
    },
  );

  tool(
    server,
    filter,
    toolsetName,
    "get_contractor",
    "Get details for a specific contractor.",
    {
      contractor_id: z.string().describe('The Check contractor ID (e.g. "ctr_xxxxx").'),
    },
    async ({ contractor_id }) => {
      const result = await checkApiGet(api, `/contractors/${contractor_id}`);
      return { content: [{ type: "text" as const, text: JSON.stringify(result, null, 2) }] };
    },
  );

  tool(
    server,
    filter,
    toolsetName,
    "create_contractor",
    "Create a new contractor.",
    {
      company: z.string().describe("The Check company ID."),
      type: z.string().optional().describe('Contractor type — "individual" or "business".'),
      first_name: z.string().optional().describe("Contractor's first name (or business primary contact's first name)."),
      middle_name: z.string().optional().describe("Contractor's middle name."),
      last_name: z.string().optional().describe("Contractor's last name (or business primary contact's last name)."),
      business_name: z.string().optional().describe("Business name (for business-type contractors)."),
      dob: z.string().optional().describe("Date of birth (YYYY-MM-DD)."),
      start_date: z.string().optional().describe("Most recent start date of contract (YYYY-MM-DD)."),
      termination_date: z.string().optional().describe("Most recent termination date (YYYY-MM-DD)."),
      workplaces: z.array(z.string()).optional().describe("List of workplace IDs where the contractor works."),
      primary_workplace: z.string().optional().describe("Workplace ID of the contractor's primary workplace."),
      email: z.string().optional().describe("Contractor's email address."),
      ssn: z.string().optional().describe("Contractor's Social Security Number."),
      ein: z.string().optional().describe("Contractor's Employer Identification Number (for businesses)."),
      default_net_pay_split: z.string().optional().describe("ID of contractor's default net pay split."),
      payment_method_preference: z.string().optional().describe('"direct_deposit" or "manual".'),
      address: z.record(z.unknown()).optional().describe("Address object with keys: line1, line2, city, state, postal_code, country."),
      metadata: z.string().optional().describe("Additional JSON metadata string."),
    },
    async ({ company, type, first_name, middle_name, last_name, business_name, dob, start_date, termination_date, workplaces, primary_workplace, email, ssn, ein, default_net_pay_split, payment_method_preference, address, metadata }) => {
      const body: Record<string, unknown> = { company };
      if (type !== undefined) body.type = type;
      if (first_name !== undefined) body.first_name = first_name;
      if (middle_name !== undefined) body.middle_name = middle_name;
      if (last_name !== undefined) body.last_name = last_name;
      if (business_name !== undefined) body.business_name = business_name;
      if (dob !== undefined) body.dob = dob;
      if (start_date !== undefined) body.start_date = start_date;
      if (termination_date !== undefined) body.termination_date = termination_date;
      if (workplaces !== undefined) body.workplaces = workplaces;
      if (primary_workplace !== undefined) body.primary_workplace = primary_workplace;
      if (email !== undefined) body.email = email;
      if (ssn !== undefined) body.ssn = ssn;
      if (ein !== undefined) body.ein = ein;
      if (default_net_pay_split !== undefined) body.default_net_pay_split = default_net_pay_split;
      if (payment_method_preference !== undefined) body.payment_method_preference = payment_method_preference;
      if (address !== undefined) body.address = address;
      if (metadata !== undefined) body.metadata = metadata;
      const result = await checkApiPost(api, "/contractors", body);
      return { content: [{ type: "text" as const, text: JSON.stringify(result, null, 2) }] };
    },
  );

  tool(
    server,
    filter,
    toolsetName,
    "update_contractor",
    "Update an existing contractor.",
    {
      contractor_id: z.string().describe("The Check contractor ID."),
      type: z.string().optional().describe('Contractor type — "individual" or "business".'),
      first_name: z.string().optional().describe("Contractor's first name."),
      middle_name: z.string().optional().describe("Contractor's middle name."),
      last_name: z.string().optional().describe("Contractor's last name."),
      business_name: z.string().optional().describe("Business name (for business-type contractors)."),
      dob: z.string().optional().describe("Date of birth (YYYY-MM-DD)."),
      start_date: z.string().optional().describe("Most recent start date of contract (YYYY-MM-DD)."),
      termination_date: z.string().optional().describe("Most recent termination date (YYYY-MM-DD)."),
      workplaces: z.array(z.string()).optional().describe("List of workplace IDs where the contractor works."),
      primary_workplace: z.string().optional().describe("Workplace ID of the contractor's primary workplace."),
      email: z.string().optional().describe("Contractor's email address."),
      ssn: z.string().optional().describe("Contractor's Social Security Number."),
      ein: z.string().optional().describe("Contractor's Employer Identification Number (for businesses)."),
      default_net_pay_split: z.string().optional().describe("ID of contractor's default net pay split."),
      payment_method_preference: z.string().optional().describe('"direct_deposit" or "manual".'),
      address: z.record(z.unknown()).optional().describe("Address object with keys: line1, line2, city, state, postal_code, country."),
      metadata: z.string().optional().describe("Additional JSON metadata string."),
    },
    async ({ contractor_id, type, first_name, middle_name, last_name, business_name, dob, start_date, termination_date, workplaces, primary_workplace, email, ssn, ein, default_net_pay_split, payment_method_preference, address, metadata }) => {
      const body: Record<string, unknown> = {};
      if (type !== undefined) body.type = type;
      if (first_name !== undefined) body.first_name = first_name;
      if (middle_name !== undefined) body.middle_name = middle_name;
      if (last_name !== undefined) body.last_name = last_name;
      if (business_name !== undefined) body.business_name = business_name;
      if (dob !== undefined) body.dob = dob;
      if (start_date !== undefined) body.start_date = start_date;
      if (termination_date !== undefined) body.termination_date = termination_date;
      if (workplaces !== undefined) body.workplaces = workplaces;
      if (primary_workplace !== undefined) body.primary_workplace = primary_workplace;
      if (email !== undefined) body.email = email;
      if (ssn !== undefined) body.ssn = ssn;
      if (ein !== undefined) body.ein = ein;
      if (default_net_pay_split !== undefined) body.default_net_pay_split = default_net_pay_split;
      if (payment_method_preference !== undefined) body.payment_method_preference = payment_method_preference;
      if (address !== undefined) body.address = address;
      if (metadata !== undefined) body.metadata = metadata;
      const result = await checkApiPatch(api, `/contractors/${contractor_id}`, body);
      return { content: [{ type: "text" as const, text: JSON.stringify(result, null, 2) }] };
    },
  );

  tool(
    server,
    filter,
    toolsetName,
    "onboard_contractor",
    "Onboard a contractor, transitioning them to active status.",
    {
      contractor_id: z.string().describe("The Check contractor ID."),
    },
    async ({ contractor_id }) => {
      const result = await checkApiPost(api, `/contractors/${contractor_id}/onboard`);
      return { content: [{ type: "text" as const, text: JSON.stringify(result, null, 2) }] };
    },
  );

  tool(
    server,
    filter,
    toolsetName,
    "list_contractor_payments_for_contractor",
    "List payments for a specific contractor.",
    {
      contractor_id: z.string().describe("The Check contractor ID."),
      limit: z.number().optional().describe("Maximum number of results to return."),
      cursor: z.string().optional().describe("Pagination cursor."),
    },
    async ({ contractor_id, limit, cursor }) => {
      const params: Record<string, unknown> = {};
      if (limit !== undefined) params.limit = limit;
      if (cursor !== undefined) params.cursor = cursor;
      const result = await checkApiList(api, `/contractors/${contractor_id}/payments`, params);
      return { content: [{ type: "text" as const, text: JSON.stringify(result, null, 2) }] };
    },
  );

  tool(
    server,
    filter,
    toolsetName,
    "get_contractor_payment_for_payroll",
    "Get a contractor payment for a specific payroll.",
    {
      contractor_id: z.string().describe("The Check contractor ID."),
      payroll_id: z.string().describe("The Check payroll ID."),
    },
    async ({ contractor_id, payroll_id }) => {
      const result = await checkApiGet(api, `/contractors/${contractor_id}/payments/${payroll_id}`);
      return { content: [{ type: "text" as const, text: JSON.stringify(result, null, 2) }] };
    },
  );

  tool(
    server,
    filter,
    toolsetName,
    "list_contractor_forms",
    "List forms for a contractor.",
    {
      contractor_id: z.string().describe("The Check contractor ID."),
      limit: z.number().optional().describe("Maximum number of results to return."),
      cursor: z.string().optional().describe("Pagination cursor."),
    },
    async ({ contractor_id, limit, cursor }) => {
      const params: Record<string, unknown> = {};
      if (limit !== undefined) params.limit = limit;
      if (cursor !== undefined) params.cursor = cursor;
      const result = await checkApiList(api, `/contractors/${contractor_id}/forms`, params);
      return { content: [{ type: "text" as const, text: JSON.stringify(result, null, 2) }] };
    },
  );

  tool(
    server,
    filter,
    toolsetName,
    "submit_contractor_form",
    "Submit a contractor form.",
    {
      contractor_id: z.string().describe("The Check contractor ID."),
      form_id: z.string().describe("The form ID."),
      parameters: z.array(z.record(z.unknown())).optional().describe('List of name/value objects representing form fields. Example: [{"name": "field_name", "value": "field_value"}].'),
    },
    async ({ contractor_id, form_id, parameters }) => {
      const body: Record<string, unknown> | undefined = parameters !== undefined ? { parameters } : undefined;
      const result = await checkApiPost(api, `/contractors/${contractor_id}/forms/${form_id}/submit`, body);
      return { content: [{ type: "text" as const, text: JSON.stringify(result, null, 2) }] };
    },
  );

  tool(
    server,
    filter,
    toolsetName,
    "reveal_contractor_ssn",
    "Reveal the SSN/EIN for a contractor.",
    {
      contractor_id: z.string().describe("The Check contractor ID."),
    },
    async ({ contractor_id }) => {
      const result = await checkApiGet(api, `/contractors/${contractor_id}/reveal`);
      return { content: [{ type: "text" as const, text: JSON.stringify(result, null, 2) }] };
    },
  );
}
